lkmpg.tex

```diff
@@ -925,6 +925,17 @@ \subsection{chardev.c}
 Don't worry if you don't see what we do with the data we read into the buffer; we don't do much with it.
 We simply read in the data and print a message acknowledging that we received it.
 
+In the multiple-threaded case, without any protection, concurrent access to the same memory may cause the race condition, and will not preserve the performance.
+In the kernel module, this problem also arises.
+So, we have to let the instance exclusively access the module.
+When we open this file, we first check \cpp|open_device_cnt| is 0 or 1.
```

Comment:
Since the meaning of variable `open_device_cnt` has been changed, it would be better to rename the variable. Something like `is_device_open` can be more precise.